Sources/DreamsNetworkInteraction.swift

```diff
@@ -0,0 +1,140 @@
+//
+//  DreamsNetworkInteraction
+//  Dreams
+//
+//  This Source Code Form is subject to the terms of the Mozilla Public
+//  License, v. 2.0. If a copy of the MPL was not distributed with this
+//  file, You can obtain one at https://mozilla.org/MPL/2.0/.
+//
+//  Copyright © 2021 Dreams AB.
+//
+
+import Foundation
+import class WebKit.WKWebView
+
+// MARK: DreamsNetworkInteracting
+
+public protocol DreamsNetworkInteracting {
+    func didLoad()
+    func use(webView: WKWebView)
+    func use(delegate: DreamsDelegate)
+    func launch(with credentials: DreamsCredentials, locale: Locale)
+}
+
+extension DreamsNetworkInteracting {
+    func launch(with credentials: DreamsCredentials) {
+        launch(with: credentials, locale: Locale.current)
+    }
+}
+
+// MARK: DreamsNetworkInteraction
+
+public final class DreamsNetworkInteraction: DreamsNetworkInteracting {
+
+    private let webService: WebServiceType
+    private let configuration: DreamsConfiguration
+    
+    private weak var webView: WKWebView!
+    private weak var delegate: DreamsDelegate?
+    
+    init(configuration: DreamsConfiguration, webService: WebServiceType) {
+        self.configuration = configuration
+        self.webService = webService
+    }
+    
+    // MARK: Public
+    
+    public func didLoad() {
+        webService.delegate = self
+        setUpUserContentController()
+    }
+    
+    public func use(webView: WKWebView) {
+        self.webView = webView
+    }
+    
+    public func use(delegate: DreamsDelegate) {
+        self.delegate = delegate
+    }
+    
+    public func launch(with credentials: DreamsCredentials, locale: Locale) {
+        loadBaseURL(credentials: credentials, locale: locale)
+    }
+    
+    // MARK: Private
+
+    private func handle(event: ResponseEvent, with jsonObject: JSONObject?) {
+        switch event {
+        case .onIdTokenDidExpire:
+            guard let requestId = jsonObject?["requestId"] as? String else { return }
+            delegate?.handleDreamsCredentialsExpired { [weak self] credentials in
+                self?.update(idToken: credentials.idToken, requestId: requestId)
+            }
+        case .onTelemetryEvent:
+            guard let name = jsonObject?["name"] as? String,
+                  let payload = jsonObject?["payload"] as? JSONObject else { return }
+            delegate?.handleDreamsTelemetryEvent(name: name, payload: payload)
+        case .onAccountProvisionRequested:
+            guard let requestId = jsonObject?["requestId"] as? String else { return }
+            delegate?.handleDreamsAccountProvisionInitiated { [weak self] in
+                self?.accountProvisionInitiated(requestId: requestId)
+            }
+        case .onExitRequested:
+            delegate?.handleExitRequest()
+        }
+    }
+    
+    private func update(idToken: String, requestId: String) {
```

Comment:
As you've moved to use `DreamsCredentials` when opening the web app, `idToken` should be `DreamsCredentials` right?
I'm not sure we need to wrap the id token with struct, I don't really see us moving away from using jwt tokens, not without bigger changes to the auth flow.

Reply:
I do not feel strongly about it. It's just cleaner _to me_ and the potential change is in arguably less significant component (struct instead of method/interface). 
To avoid changing too much I will leave it to be only an internal structure.